use domain when solving inequalities

smichr opened the pull request:
- a test for the original issue was added
- a test for the new issue raised within the same issue
  was added and other tests that were wrong were updated
- the calculation of invalid solutions was removed since
  this is already handled in the solution of the inequality

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->
closes #10671 
closes #12466
closes #10477
closes #11064
closes #12429
closes #12482
closes #12422
closes #12458

kshitij10496 commented on sympy/calculus/singularities.py:
```diff
@@ -50,13 +51,18 @@ def singularities(expression, symbol):
+        singularities = []
+        for d in denoms(expression):
+            singularities.extend(solveset(d, symbol))
+        return FiniteSet(*singularities)
```
@smichr I think that we should make a note of the fact that this function returns only the isolated singularities of the given expression. ie. we need to state that non-isolated singularities and branch points are not determined by this function.

smichr replied:
done

maurogaravello commented:
This corrects also the #12466.

maurogaravello commented:
Seems good for me.

smichr commented:
@maurogaravello , would you be willing to take a look at the commits that come after your last review?

maurogaravello commented:
I am thinking about the changes. A first impression is that making the integral test more robust is necessary. It seems to me that `assert Sum(-ln(n)/n**3, (n, 1, oo)).is_convergent() is S.true` gives a failure. The sum is clearly convergent, but `-ln(n)/n**3` is not decreasing when `n` is big. 
Mathematically speaking, the correct assumption should be the decreasing of abs(-ln(n)/n**3). 
I tried to implement this check some days ago, but sympy seems not to be able to do in all situations.

So a possibility for `Sum(expr, (n, 1, oo)).is_convergent()` could be: checking if expr is decreasing (if it is positive) or checking if -expr is decreasing (if it is negative)

smichr commented:
> but -ln(n)/n**3 is not decreasing when n is big.

True: it is increasing to 0. That is the purpose of the two tests (which essentially test `abs(f(x))`. If `f(x)` is increasing towards zero past its last maxima then `-f(x)` is *decreasing* towards 0 past the same maxima.

maurogaravello commented:
Sorry. I did not see line 494. 
I agree with your proposal.

smichr commented:
Do you want to cherry pick to your branch? I would send you a pr, but I can't right now.

smichr commented:
> remove NotImplementedError from singularities

I wonder if singularities should raise an error if anything but a finite set of expressions is obtained.

maurogaravello commented:
No problem with the PR. Your solution is better than mine.

Should I delete my PR?

smichr commented:
> Should I delete my PR?

I prefer to attribute your work to you. If you hadn't done what you did I don't know if I would have done what I did. If it's too much of a hassle we can just keep the changes here.

smichr commented:
Test is split 4 needs some debugging:

```
6149____________ sympy/integrals/tests/test_integrals.py:test_diff_wrt _____________

6150  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/integrals/tests/test_integrals.py", line 107, in test_diff_wrt

6151    assert integrate(t + 1, (t, 0, 1)) == S(3)/2

6152  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/integrals/integrals.py", line 1295, in integrate

6153    risch=risch, manual=manual)

6154  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/integrals/integrals.py", line 553, in doit

6155    evalued = Add(*others)._eval_interval(x, a, b)

6156  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/core/expr.py", line 818, in _eval_interval

6157    _singularities = singularities(self, x, include_removable=False)

6158  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/calculus/singularities.py", line 69, in singularities

6159    for d in denoms(expression, symbol):

6160  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/solvers/solvers.py", line 127, in denoms

6161    if any(s in free for s in symbols):

6162TypeError: 'Test' object is not iterable

```

`Test` is created for the purpose of the test in the suite. I'm not sure why it is being asked to iterate.

smichr commented:
And `sets` tests appear to be hanging. Also `sympy/combinatorics/tests/test_partitions.py`?

kshitij10496 commented:
> I wonder if singularities should raise an error if anything but a finite set of expressions is obtained.

I think this will restrict the use of this functionality(`singularity`) to non-trigonometric/non-periodic functions. For example, expressions with countably infinite number of disconitnuites such as`1/cos(x)` wouldn't be correctly computed by this method in the future.
What do you think about this ?

smichr commented:
> What do you think about this ?

I'm open to suggestions. I guess the easiest thing to do is just return the sympy-state-of-the-art solveset solution for the singularities and the user will make up their own mind about what to do with them. Any idea why that rational restriction was there in the first place?

smichr commented:
When tests pass, I will commit unless I hear something more from @parsoyaarihant . I removed changes to `singularities` except for the note suggested (I believe) by @kshitij10496

smichr committed:
modify singularities

    - fix XFAIL singularities tests
    - unXFAIL is_decreasing test
    - add note about what singularities omits
    - only use solveset once for is_monotonic

smichr committed:
edit calculus/util.py

    - wrap long line
    - correct error message in not_empty_in

smichr committed:
concrete/summations.py

concrete/tests/test_sums_products.py

	- modified imports
    - make integral test for convergent Sum more robust: we basically
      want to know if the expression approaches (eventually) 0 as
      infinity is approached. From the first part of the function we
      know that the upper limit is oo, so if we can find the maxima
      then we can see if the expression decreases after the last maxima.
    - move test out of XFAIL

smichr committed:
term -> factor in as_coeff_mul docstring

    This incorrect word was a carry-over from the
    as_coeff_add docstring.

smichr committed:
add _eval_interval test and unXFAIL passing test

    - test 4215 was unXFAILed

    This was fixed with pull request #12142 from
    ArifAhmed1995/issue#12133

    - a test was added to enforce that the singularites
      for which _eval_interval tests need not be rational
      (if the singularities function is used this test
      will raise an error)

        # Please enter the commit message for your changes. Lines starting
th '#' will be ignored, and an empty message aborts the commit.

smichr committed:
modify inequalities

solvers/inequalities.py
solvers/tests/test_inequalities.py

    - delay singularity calculation until NotImplementedError
      is checked
    - /!\ use `expr`, not `e`, to calculate singularities since
      in `e` terms on lhs and rhs may have cancelled
    - wrap long lines
    - only keep critical points that are within boundaries of
      interest
    - remove redundant infinities check (which should be handled
      by the loop itself)
    - factor out interval's interior point calculation to make it
      more transparent
    - update tests

smichr committed:
denoms symbols arg is now *symbols; example and tests added

smichr committed:
edit _invert docstring

smichr committed:
modify _invert_real

    - use is instead of == with Singleton
    - wrap long lines
    - dedent if/else under else condition

smichr committed:
12482: change assumptions with care

Variables were being changed to have the 'real'
assumption be False if the domain was complex
and the 'real' attribute was not already set.
But this is not implied by the complex domain.
Real -> complex but complex -x-> ~real.

smichr committed:
modify solveset

    - remove redundant use of _invalid_solutions since
      this is already handled in the function called
    - use the `domain` argument to clip the returned
      values so they are within the interval of interest
    - add tests

smichr committed:
11064: special handling of discontinuities

Singularities and solns were handled properly but
discontinuities were handled as though they were
solutions. A discontinuity might satisfy an inequality
so each discontinuity must be checked.
